### PR TITLE
Don't allow enforcing 2FA when no provider is enabled

### LINKED ISSUE #12267: Warn/check if any 2FA provider is active before enforcing 2FA
From https://github.com/nextcloud/server/issues/12249

> Another thing which confuses me is that I can enforce two factor auth even if no single two factor out method is enabled. I think it would make sense to check that at least one is enabled (u2f, totp,...) and otherwise disable the checkbox.

### PR DESCRIPTION
**Pull request by MorrisJobke**
Resubmit of #13735

Fixes #12267

I haven't committed compiled assets yet.

cc @rummatee

### CODE REVIEW

File: settings/src/store/admin-security.js [hunk 1]
```diff
@@ -33,31 +41,48 @@ export const mutations = {
...
+
+const getters = {
+	getEnabledProvidersCurrentUser(state) {
+		return state.enabledProvidersCurrentUser;
 	}
 }
 
-export const actions = {
+const actions = {
 	save ({commit}, ) {
```
Comment (skjnldsv) [inline settings/src/store/admin-security.js]: This looks weird?

File: settings/src/store/admin-security.js [hunk 2]
```diff
@@ -33,31 +41,48 @@ export const mutations = {
...
+				context.commit('setEnabledProvidersCurrentUser', response.data);
+				context.commit('stopLoading', 'providers');
+				return true;
+			})
+			.catch((error) => context.commit('API_FAILURE', error));
+	},
 }
 
-export default new Vuex.Store({
+export default {
```
Comment (skjnldsv) [inline settings/src/store/admin-security.js]: Much cleaner!! :+1:

File: settings/src/main-admin-security.js [hunk 3]
```diff
@@ -11,9 +11,10 @@ Vue.prototype.t = t;
 window.OC = window.OC || {};
 window.OC.Settings = window.OC.Settings || {};
 
-store.replaceState(
-	OCP.InitialState.loadState('settings', 'mandatory2FAState')
-)
+let initialState = OCP.InitialState.loadState('settings', 'mandatory2FAState');
+store.commit('setEnforced', initialState.enforced);
```
Comment (skjnldsv) [inline settings/src/main-admin-security.js]: This is considered bad practice to commit from a component.
You need to use actions for that.

File: settings/src/components/AdminTwoFactor.vue [hunk 4]
```diff
@@ -107,13 +117,27 @@
...
 				get: function () {
-					return this.$store.state.excludedGroups
+					return this.$store.state.security.excludedGroups
 				},
 				set: function (val) {
 					this.dirty = true
 					this.$store.commit('setExcludedGroups', val)
 				}
 			},
+			noProviderGlobally: {
```
Comment (skjnldsv) [inline settings/src/components/AdminTwoFactor.vue]: You can directly use the function if you don't have a setter here.

```suggestion
			noProviderGlobally() {
				var providers = this.$store.getters.getAllApps.filter( function(app) {
					return ('two-factor-providers' in app && 'provider' in app['two-factor-providers'] && app['active'] === true);
				});
				return (providers.length === 0);
			}
```

File: settings/src/components/AdminTwoFactor.vue [hunk 5]
```diff
@@ -78,6 +84,10 @@
 		components: {
 			Multiselect
 		},
+		beforeMount(){
+			this.$store.dispatch('getAllApps');
```
Comment (skjnldsv) [inline settings/src/components/AdminTwoFactor.vue]: Loading state here?

File: settings/src/components/AdminTwoFactor.vue [hunk 6]
```diff
@@ -78,6 +84,10 @@
...
+			this.$store.dispatch('getEnabledProvidersCurrentUser');
+		},
 		data () {
 			return {
 				loading: false,
```
Comment (skjnldsv) [inline settings/src/components/AdminTwoFactor.vue]: If so maybe set it to true by default?

**Review by skjnldsv — CHANGES_REQUESTED** [review]
A few comments :)

**Comment by MorrisJobke** [comment]
> A few comments :)

Feel free to take over - I just untangled the git history.

**Comment by rullzer** [comment]
I'm going to close this due to lack of activity.
Feel free to reopen if anybody wants to continue.

**Pull request closed without merging**